eneyj/docs/SPECIFICATION

```diff
@@ -52,231 +35,280 @@ Idea: Make keys evaluable (not implemented)
+
+4.3) Z1K2 IDs must be unique.
+     Two different ZObjects must not have the same Z1K2 ID.
+     Two ZObjects that are equal may have two different ZIDs, though.
+
+4.4) The Z1K2 ID has the form Zn or ZnKn with n natural numbers.
+
+4.5) Two ZObjects with the same Z1K2 ID must be equal.
+     This is modulo QA procedures, namespaces, caching, and metadata.
+
+4.6) A ZObject that is not persisted does not have a Z1K2 ID.
+     These are called transient ZObjects.
+
+5) Validation
+
+5.1) Every ZObject that fulfills the conditions in Section 1 is a
+     well-formed ZObject.
+
+5.2) A well-formed ZObject gets validated by running the Z4K3 validator
+     function of its Z1K1 type against itself.
+
+5.3) A validator function is a function that takes a single parameter, the
+     instance, and returns a list of Z5 errors.
 
-4.1) A Z Object may have a Z1K2 ID.
+5.4) A well-formed ZObject that returns an empty list when running the
+     validator is a valid instance of its type and a valid ZObject.
 
-4.2) A Z Object that is being persisted must have a Z1K2 ID.
+6) Evaluation
 
-4.3) The Z1K2 ID is used to refer to the Z Object.
+6.1) All Z7 function calls in a valid ZObject get evaluated.
 
-4.4) Z1K2 IDs must be unique.
-     Two different Z Objects must not have the same Z1K2 ID.
-     Two Z Objects that are equal may have two different Z IDs, though.
+6.2) The evaluation result in turn gets validated and evaluated again.
 
-x4.5) The Z1K2 ID has the form of a capital letter X, Y, or Z followed by a
-      natural decimal number.
-      It may be followed by the letter K and another natural decimal number.
+6.3) This stops when a fix point is reached, i.e. the result of the evaluation
+     is the same as its argument.
 
-4.6) Z1K2 IDs starting with Z are reserved for Wikilambda.
+6.4) Evaluating a function happens by the evaluator choosing an implementation,
+     replacing the function with that, and then executing the implementation.
 
-*4.7) Z1K2 IDs starting with Y are reserved for the team / organization working
-      together and sharing a global space.
+6.5) All arguments are passed unevaluated to the evaluator.
+     Each implementation decides when and if to evaluate any of the arguments.
+     Full evaluation may not be assumed, i.e. lazy evaluations are allowed.
 
-*4.8) Z1K2 IDs starting with X are reserved for personal definitions.
+7) Z4 Type
 
-*4.9) Z Objects in the personal space may refer to the team or global space, but
-      team or global space object must not refer to the personal space.
+7.1) Every Z1 ZObject has a Z1K1 type of type Z4.
 
-*4.10) Z Object in the team space may refer to the global space, but Z Objects
-       in the global space must not refer to the team space.
+7.2) The type system is strict and nominal.
 
-4.11) Two Z Objects with the same Z1K2 ID must be equal.
-      This is modulo QA procedures, caching, and metadata.
-      They are equal using Z33 same.
+7.3) There is no type hierarchy.
```

Comment:
I'm dubious this can be maintained, but I guess let's go with it for now. Do you want to mention Z1 here though?

Reply:
I added a sentence mentioning Z1 here.